src/main/io/gimbal_control.c

```diff
@@ -0,0 +1,326 @@
+
+    gimbalCmdOut.crc = crc;
+
+    return true;
+}
+
+// Gimbal updates should be sent at 100Hz or the gimbal will self center after approx. 2 seconds
+void gimbalUpdate(timeUs_t currentTimeUs)
+{
+    static enum {GIMBAL_OP1, GIMBAL_OP2, GIMBAL_CMD} gimbalParseState = GIMBAL_OP1;
+    static timeUs_t lastRxTimeUs = 0;
+
+    if (!gimbalSerialPort) {
+        setTaskEnabled(TASK_GIMBAL, false);
+        return;
+    }
+
+    // Read bytes from the VTX gimbal serial data stream
+
+    uint32_t bytes =  serialRxBytesWaiting(gimbalSerialPort);
+
+    if (bytes > 0) {
+        lastRxTimeUs = currentTimeUs;
+        while (bytes--) {
+            uint8_t inData = serialRead(gimbalSerialPort);
+
+            // If the packet is a gimbalCmd_t structure then parse, otherwise pass through
+            switch (gimbalParseState) {
+            default:
+            case GIMBAL_OP1:
+                if (inData == (GIMBAL_CMD_S & 0xff)) {
+                    gimbalParseState = GIMBAL_OP2;
+                } else {
+                    serialWrite(gimbalSerialPort, inData);
+                }
+                break;
+            case GIMBAL_OP2:
+                if (inData == ((GIMBAL_CMD_S >> 8) & 0xff)) {
+                    gimbalParseState = GIMBAL_CMD;
+                    gimbalInCount = sizeof(gimbalCmdIn.u.gimbalCmd.opcode);
+                } else {
+                    serialWrite(gimbalSerialPort, GIMBAL_CMD_S && 0xff);
+                    serialWrite(gimbalSerialPort, inData);
+                    gimbalParseState = GIMBAL_OP1;
+                }
+                break;
+            case GIMBAL_CMD:
+                gimbalCmdIn.u.bytes[gimbalInCount++] = inData;
+                if (gimbalInCount == sizeof(gimbalCmdIn.u.gimbalCmd)) {
+                    gimbalCmdOut = gimbalCmdIn.u.gimbalCmd;
+                    gimbalSet(gimbalCmdIn.u.gimbalCmd.roll, gimbalCmdIn.u.gimbalCmd.pitch, gimbalCmdIn.u.gimbalCmd.yaw);
+                    serialWriteBuf(gimbalSerialPort, (uint8_t *)&gimbalCmdOut, sizeof(gimbalCmdOut));
+                    gimbalParseState = GIMBAL_OP1;
+                }
+                break;
+            }
+        }
+    } else if (cmpTimeUs(currentTimeUs, lastRxTimeUs) > GIMBAL_HT_TIMEOUT_US) {
+        gimbalSet(0, 0, 0);
+        serialWriteBuf(gimbalSerialPort, (uint8_t *)&gimbalCmdOut, sizeof(gimbalCmdOut));
+    }
+}
+
+bool gimbalInit(void)
+{
+    const serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_GIMBAL);
+    if (!portConfig) {
+        return false;
+    }
+
+    // Serial communications is 115200 8N1
+    gimbalSerialPort = openSerialPort(portConfig->identifier, FUNCTION_GIMBAL,
+                                      NULL, NULL,
+                                      115200, MODE_RXTX, SERIAL_STOPBITS_1 | SERIAL_PARITY_NO);
+
+    gimbalCmdIn.u.gimbalCmd.opcode = GIMBAL_CMD_S;
+    gimbalCmdOut.opcode = GIMBAL_CMD_S;
```

Comment:
This is just microoptimization, you can set additional few bytes when constructing packet

Reply:
Is this really a bad thing?

Reply:
It is slightly faster, but more fragile. Some future PR may easily break it.